src/lib/cache_lib/impl.ml

```diff
@@ -0,0 +1,236 @@
+
+    let name {name; _} = name
+
+    let create (type elt) ~name ~logger
+        (module Elt : Hash_set.Elt_plain with type t = elt) : elt t =
+      let set =
+        Hash_set.create ~growth_allowed:true ?size:None (module Elt) ()
+      in
+      {name; set; logger}
+
+    let register t x =
+      let%map () = Hash_set.strict_add t.set x in
+      Cached.create ~logger:t.logger t x
+
+    let mem t x = Hash_set.mem t.set x
+
+    let remove t x = Hash_set.strict_remove t.set x
+  end
+  
+  and Cached : sig
+    include Intf.Cached.S
+
+    val create : logger:Logger.t -> 'elt Cache.t -> 'elt -> ('elt, 'elt) t
+  end = struct
+    type (_, _) t =
+      | Base :
+          { data: 'a
+          ; cache: 'a Cache.t
+          ; mutable consumed: bool }
+          -> ('a, 'a) t
+      | Derivative :
+          { original: 'a
+          ; mutant: 'b
+          ; parent: ('c, 'a) t
+          ; mutable consumed: bool }
+          -> ('b, 'a) t
+      | Phantom : 'a -> ('a, _) t
+
+    let phantom x = Phantom x
+
+    let rec cache : type a b. (a, b) t -> b Cache.t = function
+      | Base x -> x.cache
+      | Derivative x -> cache x.parent
+      | Phantom _ -> failwith "cannot access cache of phantom Cached.t"
+
+    let value : type a b. (a, b) t -> a = function
+      | Base x -> x.data
+      | Derivative x -> x.mutant
+      | Phantom x -> x
+
+    let original : type a b. (a, b) t -> b = function
+      | Base x -> x.data
+      | Derivative x -> x.original
+      | Phantom _ -> failwith "cannot access original of phantom Cached.t"
+
+    let was_consumed : type a b. (a, b) t -> bool = function
+      | Base x -> x.consumed
+      | Derivative x -> x.consumed
+      | Phantom _ ->
+          failwith "cannot determine consumption of phantom Cached.t"
+
+    let mark_consumed : type a b. (a, b) t -> unit = function
+      | Base x -> x.consumed <- true
+      | Derivative x -> x.consumed <- true
+      | Phantom _ -> failwith "cannot set consumption of phantom Cached.t"
+
+    let create ~logger cache data =
+      let t = Base {data; cache; consumed= false} in
+      Gc.Expert.add_finalizer (Heap_block.create_exn t) (fun block ->
+          let t = Heap_block.value block in
+          if not (was_consumed t) then
+            Inputs.handle_unconsumed_cache_item ~logger
+              ~cache_name:(Cache.name cache) ) ;
+      t
+
+    let assert_not_consumed t msg =
+      let open Error in
+      if was_consumed t then raise (of_string msg)
+
+    let peek (type a b) (t : (a, b) t) : a =
+      assert_not_consumed t "cannot peek at consumed Cached.t" ;
+      value t
+
+    let consume (type a b) (t : (a, b) t) : unit =
+      assert_not_consumed t "cannot consume Cached.t twice" ;
+      mark_consumed t
+
+    let transform (type a b) (t : (a, b) t) ~(f : a -> 'c) : ('c, b) t =
+      consume t ;
+      Derivative
+        {original= original t; mutant= f (value t); parent= t; consumed= false}
```

Comment:
Should you also add a finalizer on the derivation? Without that, we could still be forgetting to consume derived cached values.

Reply:
good catch